R/civis_ml_utils.R

```diff
@@ -19,19 +19,93 @@ CIVIS_ML_CLASSIFIERS <- c("sparse_logistic",
-  stringsAsFactors = FALSE
-)
+
+get_template_ids_all_versions <- function(){
+  # Get template IDs for all accessible CivisML versions.
+
+  template_alias_objects <- fetch_until(aliases_list,
+                                       limit=1000,
+                                       object_type='template_script',
+                                       function(x) length(x) == 0)
+
+
+  template_aliases <- sapply(template_alias_objects, `[[`, 'alias')
+
+  # Template script aliases are not used exclusively for CivisML
+  civisml_template_alias_objects <- template_alias_objects[startsWith(template_aliases,
+                                                                      'civis-civisml-')]
+
+  ids <- lapply(civisml_template_alias_objects, function(alias_obj){
+
+        id = alias_obj$objectId
+
+        job_type_version = get_job_type_version(alias_obj$alias)
+
+        return(list(id = id,
+                    version = job_type_version$version,
+                    name = job_type_version$job_type))
+
+    })
+
+  CIVIS_ML_TEMPLATE_IDS <- data.frame(matrix(unlist(ids),
+                                  nrow=length(ids),
+                                  byrow=T),
+                           stringsAsFactors=FALSE)
+
+  names(CIVIS_ML_TEMPLATE_IDS) <- c("id", "version", "name")
+  CIVIS_ML_TEMPLATE_IDS[,"id"] <- as.integer(CIVIS_ML_TEMPLATE_IDS[,"id"])
+
+
+  return(CIVIS_ML_TEMPLATE_IDS)
+
+}
+
+get_job_type_version <- function(alias){
+  # Derive the job type and version from the given alias.
+
+  # `alias` is a character vector and length(`alias`) = 1
+
+  # A version-less alias for production, e.g., "civis-civisml-training"
+  match_production <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)\\Z', alias, perl=TRUE)))
+  # A versioned alias, e.g., "civis-civisml-training-v2-3"
+  match_v <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)-v(\\d+)-(\\d+)\\Z', alias, perl=TRUE)))
+  # A special-version alias, e.g., "civis-civisml-training-dev"
+  match_special <- unlist(regmatches(alias, regexec('\\Acivis-civisml-(\\w+)-(\\S+[^-])\\Z', alias, perl=TRUE)))
+
+
+  if (!identical(match_production, character(0))) {
+
+    job_type = match_production[2]
+    version = "prod"
+
+  } else if (!identical(match_v, character(0))) {
+
+    job_type = match_v[2]
+    version = paste0("v",match_v[3],".",match_v[4])
+
+  } else if (!identical(match_special, character(0))) {
+
+    job_type = match_special[2]
+    version = match_special[3]
+
+  } else {
+
+    stop(paste('Unable to parse the job type and version from the CivisML alias:', alias))
+
+  }
+
+  return(list(job_type = job_type, version = version))
+
+}
 
 # returns a version compatible template id for a given training model without API calls.
 get_predict_template_id <- function(m) {
   train_id <- m$job$fromTemplateId
   this_version <- CIVIS_ML_TEMPLATE_IDS[CIVIS_ML_TEMPLATE_IDS$id == train_id, "version"]
```

Comment:
If the module-level constant `CIVIS_ML_TEMPLATE_IDS` is removed above, how is this line still operating without crashing while trying to access `CIVIS_ML_TEMPLATE_IDS`? What do I miss?

Reply:
I would imagine this is crashing : )

This should be refactored to fetch or re-use pre-fetched civis ml template ids.

Reply:
Thanks for calling this out! I added a line to fetch the civis ml template ids.